File: CsChatClient/Models/UserInfo.cs
using System;
using Newtonsoft.Json;

namespace CsChatClient.Models
{
    public class UserInfo : ISerializable
    {
        public string Uid { get; set; }
        public string Name { get; set; }
        public string DisplayName { get; set; }

        /// <summary>
        /// 用户在群里面给自己设置的备注，不同群不一样
        /// </summary>
        public string GroupAlias { get; set; }

        /// <summary>
        /// 我为好友设置的备注
        /// </summary>
        public string FriendAlias { get; set; }

        public string Portrait { get; set; }
        public int Gender { get; set; }
        public string Mobile { get; set; }
        public string Email { get; set; }
        public string Address { get; set; }
        public string Company { get; set; }
        public string Social { get; set; }
        public string Extra { get; set; }
        public long UpdateDt { get; set; }

        /// <summary>
        /// 0 normal; 1 robot; 2 thing;
        /// </summary>
        public int Type { get; set; }

        /// <summary>
        /// normal; 1 deleted;
        /// </summary>
        public int Deleted { get; set; }

        public void Serialize(JsonWriter writer)
        {
            throw new NotImplementedException();
        }

        public bool Unserialize(JsonReader reader)
        {
            while (reader.Read())
            {
                switch (reader.TokenType)
                {
                    case JsonToken.PropertyName:
                        if (reader.Value.Equals("gender"))
                        {
                            Gender = JsonTools.GetNextInt(reader);
                        }
                        else if (reader.Value.Equals("type"))
                        {
                            Type = JsonTools.GetNextInt(reader);
                        }
                        else if (reader.Value.Equals("deleted"))
                        {
                            Deleted = JsonTools.GetNextInt(reader);
                        }
                        else if (reader.Value.Equals("updateDt"))
                        {
                            UpdateDt = JsonTools.GetNextBigInt(reader);
                        }
                        else if (reader.Value.Equals("uid"))
                        {
                            Uid = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("name"))
                        {
                            Name = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("displayName"))
                        {
                            DisplayName = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("groupAlias"))
                        {
                            GroupAlias = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("friendAlias"))
                        {
                            FriendAlias = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("portrait"))
                        {
                            Portrait = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("mobile"))
                        {
                            Mobile = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("email"))
                        {
                            Email = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("address"))
                        {
                            Address = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("company"))
                        {
                            Company = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("social"))
                        {
                            Social = JsonTools.GetNextString(reader);
                        }
                        else if (reader.Value.Equals("extra"))
                        {
                            Extra = JsonTools.GetNextString(reader);
                        }
                        else
                        {
                        }

                        break;
                    case JsonToken.EndObject:
                        return true;
                }
            }

            return false;
        }
    }
}

File: CppChatClient/model/UserInfo.cpp
//
//  UserInfo.cpp
//  ChatClient
//
//  Created by Hao Jia on 2020/1/30.
//  Copyright © 2020 WF Chat. All rights reserved.
//

#include "UserInfo.h"
#include "../JsonTools.h"
    
namespace WFCLib {
void UserInfo::Unserialize(const void *pvalue) {
    const Value& value = *((const Value *)pvalue);
    if (!value.IsObject()) {
        return;
    }

//            std::string uid;
    getValue(value, "uid", uid);
//            std::string name;
    getValue(value, "name", name);
//            std::string displayName;
    getValue(value, "displayName", displayName);
//            int gender;
    getValue(value, "gender", gender);
//            std::string portrait;
    getValue(value, "portrait", portrait);
//            std::string mobile;
    getValue(value, "mobile", mobile);
//            std::string email;
    getValue(value, "email", email);
//            std::string address;
    getValue(value, "address", address);
//            std::string company;
    getValue(value, "company", company);
//            std::string social;
    getValue(value, "social", social);
//            std::string extra;
    getValue(value, "extra", extra);
//            //0 normal; 1 robot; 2 thing;
//            int type;
    getValue(value, "type", type);
//            int64_t updateDt;
    getValue(value, "updateDt", updateDt);
    getValue(value, "friendAlias", friendAlias);
    getValue(value, "groupAlias", groupAlias);
    getValue(value, "deleted", deleted);
}

void UserInfo::Serialize(void *pwriter) const {
    Writer<StringBuffer> &writer = *((Writer<StringBuffer> *)pwriter);

    writer.StartObject();

//            std::string uid;
    writer.String("uid");
    writer.String(uid);
//            std::string name;
    writer.String("name");
    writer.String(name);
//            std::string displayName;
    writer.String("displayName");
    writer.String(displayName);
//            int gender;
    writer.String("gender");
    writer.Int(gender);
//            std::string portrait;
    writer.String("portrait");
    writer.String(portrait);
//            std::string mobile;
    writer.String("mobile");
    writer.String(mobile);
//            std::string email;
    writer.String("email");
    writer.String(email);
//            std::string address;
    writer.String("address");
    writer.String(address);
//            std::string company;
    writer.String("company");
    writer.String(company);
//            std::string social;
    writer.String("social");
    writer.String(social);
//            std::string extra;
    writer.String("extra");
    writer.String(extra);
    
    writer.String("friendAlias");
    writer.String(friendAlias);
    
    writer.String("groupAlias");
    writer.String(groupAlias);
//            //0 normal; 1 robot; 2 thing;
//            int type;
    writer.String("type");
    writer.Int(type);
//            int64_t updateDt;
    writer.String("updateDt");
    writer.Int64(updateDt);
    
    writer.String("deleted");
    writer.Int(deleted);

    writer.EndObject();
}
};
